034-instanced-pipelines/proposal.md

```diff
@@ -0,0 +1,74 @@
+# Summary
+
+Instanced pipelines group together pipelines which share a common template configured with different ((vars)). They provide a simple two-level hierarchy and automatic archiving of instances which are no longer needed.
+
+# Proposal
+
+Pipelines can be configured with 'instance vars' like so:
+
+```yaml
+fly set-pipeline -p branch --instance-var branch=feature/foo
+```
+
+This will configure a `branch` pipeline, with the `((branch))` var set to `"feature/foo"`.
+
+Instance vars are used as part of the pipeline identifier in the UI and API. There can be multiple instances of a pipeline with the same name:
+
+```yaml
+fly set-pipeline -p branch --instance-var branch=feature/foo
+fly set-pipeline -p branch --instance-var branch=feature/bar
+```
+
+Instanced pipelines sharing the same name will be grouped together in the web UI.
+
+An individual instance of a pipeline can be manually destroyed, paused, and archived ([RFC #33](https://github.com/concourse/rfcs/pull/33)):
+
+```yaml
+fly destroy-pipeline -p branch -i branch:feature/foo
+fly pause-pipeline   -p branch -i branch:feature/foo
+fly archive-pipeline -p branch -i branch:feature/foo
+```
+
+(Side note: `:` vs. `=` is a little weird but it's consistent with `fly check-resource` - we use `=` for assignment and `:` for partial filtering.)
+
+## Automatic archival
+
+Instanced pipelines build on the automatic pipeline archiving introduced in [RFC #33](https://github.com/concourse/rfcs/pull/33). Individual pipeline instances that are no longer configured will be automatically archived in the same way that normal pipelines would.
+
+For example, say I have a job whose build plan configures a pipeline instance for each supported version:
+
+```yaml
+plan:
+- get: ci
+- set_pipeline: release
+  file: ci/pipelines/release.yml
+  instance_vars:
+    version: 5.3
+- set_pipeline: release
+  file: ci/pipelines/release.yml
```

Comment:
How about if I use a different pipeline yml here? Is that allowed? 

- If yes, then this feature is just a way to group pipelines with a single name. In that case, the existing `var` and `var_file` is good enough, no need to introduce `instance_vars`.
- If not, then why let users to input `file` multiple times? The syntax may change to something like:

```yaml
- set_pipeline: release
  file: ci/pipelines/release.yml
  instance_vars:
  - version: 5.3
    foo: bar
  - version: 5.2
    foo: bar
```

Reply:
The `set_pipeline` step sets all instances of a pipeline at once. They don't technically have to share a common template, that's just how most examples are framed as that seems like the most obvious use case.

Your second example would be accomplished by way of the `across` step outlined in #29:

```yaml
across:
- from: 3.0
  to: 5.0
- from: 4.0
  to: 5.0
as: upgrade_path
do:
- set_pipeline: upgrade-test
   file: ci/pipelines/upgrade-test.yml
  instance_vars:
    upgrade_from: ((upgrade_path.from))
    upgrade_to: ((upgrade_path.to))
```

...though I'm currently leaning towards this syntax instead, since it can be done by just adding the `matrix:` field to any step rather than the full `across:/as:/do:`:

```yaml
set_pipeline: upgrade-test
file: ci/pipelines/upgrade-test.yml
instance_vars:
  upgrade_from: ((.:upgrade_path.from))
  upgrade_to: ((.:upgrade_path.to))
matrix:
  upgrade_path:
  - from: 3.0
    to: 5.0
  - from: 4.0
    to: 5.0
```

In principle, we could make the `file:` field also parameterized through of the matrix if that sort of thing is necessary:

```yaml
set_pipeline: upgrade-test
file: ci/pipelines/((.:upgrade_path.file))
instance_vars:
  upgrade_from: ((.:upgrade_path.from))
  upgrade_to: ((.:upgrade_path.to))
matrix:
  upgrade_path:
  - from: 3.0
    to: 5.0
    file: old-upgrade-test.yml
  - from: 4.0
    to: 5.0
    file: new-upgrade-test.yml
```

So for that case there's probably merit in #5277 after all, though I still don't think we should allow the name to be interpolated as it really overlaps with the intended function of instanced pipeline names; `instance_vars` effectively become part of the identifier, and they don't require you to sanitize whatever you're interpolating since the values aren't part of the pipeline name and thus don't have to conform to #40.